### ISSUE TITLE
[Pallas:TPU:SparseCore] JAX-produced lookup buffer causes vld_address_out_of_range in consumer

### ISSUE THREAD

**Issue by samixyzdev** [jax-ml/jax#39662]
### Description

On a TPU v6e, I can run a Pallas SparseCore consumer successfully when its
lookup-table input is a resident/static device buffer. However, the same
consumer fails when an ordinary JAX computation builds an equivalent buffer
online and passes it to the SparseCore kernel.

The producer uses regular JAX scatter/update operations. The resulting array
has the same logical shape, dtype, contents, padding, and requested device
layout as the working resident input. The producer output also matches a CPU
reference exactly before it is passed to the consumer.

The SparseCore kernel uses:

```python
pltpu.CompilerParams(needs_layout_passes=False)
```

The failure is a device-side halt such as:

```text
vld_address_out_of_range
```

Some variants also report:

```text
result_fifo_not_empty_at_task_commit
```

### Expected behavior

An array produced by an ordinary JAX computation should be consumable by a
Pallas SparseCore kernel when its shape, dtype, padding, and declared layout
match those of a resident input accepted by the same kernel.

If this producer-to-SparseCore handoff is unsupported, compilation should
reject it with a clear diagnostic instead of allowing a device-side
out-of-range load.

### Observed behavior matrix

| Producer | Consumer | Result |
|---|---|---|
| Resident/static device buffer | Pallas SparseCore | Works |
| Ordinary JAX scatter/update | Ordinary JAX lookup and aggregation | Works; exact result |
| Ordinary JAX scatter/update | Pallas SparseCore | Device halt |
| Ordinary JAX scatter/update | A separate SparseCore workload with a similar 2-D buffer | Works |

This suggests that SparseCore execution itself and ordinary JAX construction
are both functional, but one particular producer/consumer handoff or layout
combination is not.

### Variants already tested

The failure remains after trying:

1. Blocking on the JAX producer before launching the SparseCore consumer.
2. Wrapping the producer and consumer in one outer `jax.jit`.
3. Copying into a fresh device buffer with the same requested JAX layout.
4. Writing into a donated, preallocated workspace and verifying buffer reuse.
5. A 2-D SparseCore-friendly layout with a 128-element minor dimension.
6. Explicit physical padding to multiples of `8 x 128`.
7. Passing each small lookup table independently.
8. Packing all lookup tables into one contiguous buffer.
9. Building the packed representation directly in the donated workspace.
10. Fused and staged consumer variants.

All online producer outputs were checked against a CPU reference before the
SparseCore launch.

### Environment

- Device: TPU v6e
- JAX: 0.10.2
- `jaxlib`: 0.10.2
- `libtpu`: 0.0.42.1
- Python: 3.11.0rc1

### Reproduction

The current reproducer is part of a larger analytical-query experiment. I am
reducing it to standalone synthetic arrays containing one ordinary JAX scatter
producer, one resident-buffer control, and one minimal SparseCore load/probe
consumer. I can post that reproduction in a follow-up comment.

**Comment by samixyzdev** [comment]
Follow-up on the standalone reduction I promised above:

I reduced the producer/consumer handoff to synthetic arrays and tested it on
the same TPU v6e environment (JAX/jaxlib 0.10.2, libtpu 0.0.42.1). The reduced
cases **do not reproduce the failure**.

The following all execute correctly and match NumPy references:

- a JAX scatter-set buffer passed to a SparseCore `sync_copy` + 8-lane
  `plsc.load_gather` consumer;
- the producer and consumer composed under one outer `jax.jit`;
- three packed lookup tables sliced into separate VMEM buffers;
- an SF-sized packed buffer;
- 300,000 colliding JAX scatter-add updates;
- 1.45M SparseCore probes distributed with `emit_pipeline`;
- SparseCore lookup followed by VMEM-local survivor compaction and SparseCore
  scatter-add aggregation.

The resident-buffer controls and JAX-produced-buffer cases both pass in all of
these reductions.

This means my original attribution to a generic "ordinary JAX scatter output
-> SparseCore" handoff was too broad. The full composition still reproducibly
halts with `vld_address_out_of_range`, but the trigger requires some additional
detail that is absent from these reductions. I am continuing a delta reduction
of the full consumer rather than posting a misleading reproducer. I will follow
up with a failing standalone script once that missing condition is isolated.

**Comment by apaszke** [comment]
In some cases, if you use unexpected shapes and run with layout passes disabled, we are aware that miscompiles can happen. We are currently working on enabling layout passes by default for all workloads and removing the option to opt out, which would patch up that hole. I strongly suspect that the error has nothing to do with a dynamic JAX array feeding a kernel and is simply an artifact of an unexpected behavior of the pass pipeline with layout passes disabled.

If there's any particular op that is blocking you from using layout passes please open another issue. If you find a reproducer for this, please reopen. Otherwise, there's nothing we can do without a way to reproduce your problem. Sorry!

**Comment by samixyzdev** [comment]
Update after retesting on a fresh TPU v6e VM with the same software versions (JAX/jaxlib 0.10.2 and libtpu 0.0.42.1):\n\nThe current full producer/consumer composition no longer reproduces the device halt. I added a diagnostic switch that bypasses the canonical-buffer copy and passes the ordinary JAX-built lookup tables directly to the SparseCore consumer. All of the following now execute successfully and match the CPU references:\n\n- SF1: Q4.1, Q4.2, and Q4.3;\n- SF5: Q4.1, Q4.2, and Q4.3;\n- fused, chunked, and staged SparseCore plans;\n- direct JAX-builder output -> SparseCore consumer, without an intermediate canonical copy.\n\nThe previously posted standalone synthetic reductions also continue to pass. Therefore I still do not have a failing standalone reproducer, and I will not reopen the issue based on the earlier attribution. The original failure likely depended on an implementation detail in the earlier full composition that is no longer present. I will keep the direct-handoff case as a regression test and follow up only if I can isolate a genuinely failing case.
